Do final alignment based on presence of <,>,^,0= in format string

agroener opened the pull request:
Beginning to finalize the table-column alignment options for pprint in Table package.

mhvk commented:
@agroener - just to check whether you're able to rebase this. If not, one of us should be able to do it.

agroener commented:
@mhvk I'll give it a shot right now.

agroener commented:
While travis is working, it seems to be complaining about 'col_width' not being defined before being called. Is this a value which has been done away with?

agroener commented:
Running into an issue when trying to import astropy:

    WARNING: You appear to be trying to import astropy from within a source checkout without building 
    the extension modules first.  Attempting to (re)build extension modules: [astropy]
    Rebuilding extension modules [Done]
    ERROR: An error occurred while attempting to rebuild the extension modules.  Please try manually 
    running `./setup.py develop` or `./setup.py build_ext --inplace` to see what the issue was.  Extension 
    modules must be successfully compiled and importable in order to import astropy. [astropy]

I made sure to do:

    python setup.py develop
before I imported. Any ideas?

mhvk commented:
@agroener - One problem is that you should never merge `upstream/master` in, but rather you need to
```
git rebase upstream/master
```
(Don't know if the problems with `develop` are related or not.)

agroener commented:
Okay - I rebased onto upstream/master. Am I allowed to ask why I should never merge upstream/master?

embray commented:
> Am I allowed to ask why I should never merge upstream/master?

The short version (the long version involves drawings and jargon) is that when your pull request--i.e. your branch--is merged it will be merged back into master.  So if you've doing bidirectional merges from master to your branch, and later from your branch back into master it makes an unholy mess of the commit history in master.

embray commented on CHANGES.rst:
```diff
@@ -543,6 +543,8 @@ New Features
 - ``astropy.tests``
 
   - Added a new Quantity-aware ``assert_quantity_allclose``. [#3273]
```
Is this going to be for v1.1.0?  If so this should be moved to the right section.

taldcroft commented:
I should have time to take a look at this on Friday.

taldcroft commented:
@agroener - this is failing lots of tests.  Please make sure your PR is passing tests locally (preferably by testing for both Python 2.7 and 3.4 prior to pushing).

taldcroft commented on astropy/table/pprint.py:
```diff
@@ -185,7 +185,7 @@ def _get_pprint_size(max_lines=None, max_width=None):
 
     def _pformat_col(self, col, max_lines=None, show_name=True, show_unit=None,
-                     show_dtype=False, show_length=None, html=False):
+                     show_dtype=False, show_length=None, html=False, align='right'):
```
I think this is a leftover from an earlier version since it doesn't appear that `align` is used anywhere in this function.

taldcroft replied:
Never mind the previous comment about the `align` keyword (which I have now deleted).  I think there *should* be an `align` keyword here, and it should override the existing align if that is found in the `col.format`.  The `align` keyword should also be added to the public `Column.pformat` method.

taldcroft commented:
I'm wondering about making it easier to use the `align` keyword for multiple columns.  Right now you can suppy a list of `right`, `left` or `center` strings, but for more than 2 or 3 columns this gets pretty cumbersome.  What about just being consistent with the `format` syntax and using `< ^ >` instead and saying that `align` is always just a string.  So e.g.
```
t.pprint(align='^ ^ ^ < < =0 < >')  # specify align for all columns (whitespace is insignificant)
t.pprint(align='<')  # left align for all columns
```
FWIW this reduces the English bias by changing words into symbols.  As mentioned we can have whitespace be optional but encouraged, so `align='^^>'` would be allowed.

agroener commented:
Okay - there are a lot of error from Travis which I have no idea how to interpret. They're quite cryptic...

I'm also having an issue with building/importing astropy locally.

    WARNING: You appear to be trying to import astropy from within a source checkout without building 
    the extension modules first.  Attempting to (re)build extension modules: [astropy]
    Rebuilding extension modules [Done]
    ERROR: An error occurred while attempting to rebuild the extension modules.  Please try manually 
    running `./setup.py develop` or `./setup.py build_ext --inplace` to see what the issue was.  Extension 
    modules must be successfully compiled and importable in order to import astropy. [astropy]

Does anyone have any ideas on how to fix this? I've spent some time trying to look this one up, but it seems to be an astropy specific thing, so I'm not finding any solutions online.

bsipocz commented:
@agroener - Try to run 
```python setup.py develop```

After that  the testing env should work, I think.
```python setup.py test```

agroener commented:
Hi @bsipocz , this is actually the error which I obtain after running that command. (See above). For some reason, I'm still having this issue.

taldcroft commented:
@agroener - have you first tried cleaning your repo with `git clean -fxd` and then rebuilding?  *Warning*, this removes all non-tracked files so if you made some of your own new files in the astropy directory don't do that command.

And what sort of python installation are you using for installing in to?  For astropy development I recommend using Anaconda to have a clean environment.

agroener commented:
@taldcroft I'm using: 
Python 2.7.9 |Anaconda 2.1.0

I don't believe I have any of my own files in this repo, just modified versions of pprint.py and such. I'll give it a shot and see if it works.

bsipocz commented:
@agroener - testing is not working in ipython at the moment. 
Though it's weird why the import isn,t working in ipython, it was usually the other way around for me (but that was a totally different situation, messed up PATH/PYTHONPATH due to some aliases I defined in .bashrc). 
I hope some of the gurus can help you to sort this out.

agroener commented:
Okay - looking through some of the output of testing, I see that a lot of tests fail in test_pprint.py. I think the alignment feature may have something to do with this since by default I am using 'right' as my default alignment.

@bsipocz Thanks for the infor, and previously the suggestion. I'm okay without ipython for now, but I agree, it is weird I'm not even able to import the library.

embray commented:
Are you using virtualenv?

agroener commented:
@embray I tried to get it working that way a while back but gave up on it. So no, I do not believe I am using virtualenv (if somehow it is being used, it is not to my knowledge).

embray commented:
I'm not sure what `./setup.py develop` does in anaconda, unless you're using a specific conda env for that.

embray commented:
For the most part you don't even need that though, as long as your cwd is the root of the Astropy source repository.  `./setup.py develop` is only helpful if you want to create a semi-permanent link to your development source on your python path.

mhvk commented:
@agroener - I ran some tests locally and the problem is that the column headings also get aligned, not just the column content. This should be relatively easy to fix...

agroener commented:
@mhvk I'll take a crack at this today. I also noticed that 'col_strs' is being appended to twice - clearly I have some redundant code in here. I think the cause of this is that there have been upgrades in the midst of this PR, which I've pulled in.

agroener commented:
Actually, regarding the double appending of 'col_strs' - it looks like it is coming from the generator '_pformat_col_iter.' So I'm not entirely sure how or why this is happening. I just know it is already incorrect coming into the '_pformat_col'. Any ideas why this is happening?

agroener commented:
Looks like this last commit has fixed everything. Waiting on tests to confirm.

taldcroft commented:
@agroener - this is still failing doctests.  Are you getting a clean run of `python setup.py test --package=table` locally?

About the list of contributors, I think that gets done by the release manager prior to a new release.

astrofrog commented:
(Just to confirm, the list of contributors does get done by the package manager before a release)

agroener commented:
@taldcroft Didn't know you had the option to test a single package - very handy :)

Just ran it and this is the output:
    2 failed, 1588 passed, 3 xfailed in 36.15 seconds

And it looks like pprint is passing:
    astropy/table/tests/test_pprint.py ............................................................................................................

Any suggestions about how to fix those failed doctests? I seem to remember Travis complaining about spaces and poorly formatted docstrings, etc.

taldcroft commented:
@agroener - you should be able to copy/paste the docstring examples interactively in IPython and (presumably) reproduce the issue and debug.  If you can copy/paste all the examples without getting any errors then report back on that.

agroener commented:
Alright - After fixing a few things, I have all of the tests passing locally:

    1590 passed, 3 xfailed in 34.85 seconds 

Here is hoping everything works remotely. I would also appreciate any/all comments on how things look or work. I've updated the docs a bit, so as to include an example of how one can perform alignment of columns in either of the two ways I have implemented it.

agroener commented:
1 sphinx error remaining. I've looked at the output of Travis and I'm at a loss for where this error is occuring. Any ideas?

bsipocz commented on docs/table/access_table.rst:
```diff
@@ -374,6 +374,49 @@ For columns the syntax and behavior of
   Length = 100 rows
 
+  Column alignment
+''''''''''''''''
```
The extra spaces before "Column alignment" are the culprit.

The log gives the hint, here: 
```
/home/travis/build/astropy/astropy/docs/table/access_table.rst:378: WARNING: Literal block ends without a blank line; unexpected unindent.
```

agroener replied:
@bsipocz Ah, awesome! Thank you so much. I also have a very difficult time finding these kinds of errors :)

bsipocz replied:
@agroener - Sure, no problem. I stared a lot of sphinx outputs when the docs was still full of build warnings, so now know what to look for :)

bsipocz commented on astropy/table/table.py:
```diff
@@ -588,7 +588,8 @@ def _base_repr_(self, html=False):
             tableid=tableid, html=html, max_width=(-1 if html else None),
-            show_name=True, show_unit=None, show_dtype=True)
+                                                    show_name=True, show_unit=None,
+                                                         show_dtype=True)
```
this is just nitpicking: as the indentation of 'tableid' was brought back to the beginning of the line (rather than under 'self'), this should be aligned with that one, too (and thus you don't need two lines to fit in the 80 char line length)

mhvk replied:
@agroener - I think you might as well reformat this while getting rid of the sphinx error (@bsipocz - thanks for finding it; I always have trouble with the sphinx output, trying to understand errors!).  I'd do
```
data_lines, outs = self.formatter._pformat_table(
    self, tableid=tableid, html=html, max_width=(-1 if html else None),
    show_name=True, show_unit=None, show_dtype=True)
```

agroener replied:
Okay - I've went ahead and made that change. Thanks.

taldcroft commented:
@mhvk @agroener - what do you think about the idea I raised earlier of making the API a bit simpler and more consistent by only allowing a single string which re-uses the characters from `format`:
```
t.pprint(align='^ ^ ^ < < =0 < >')  # specify align for all columns (whitespace is insignificant)
t.pprint(align='<')  # left align for all columns
```
I think this could also simplify the code logic a bit.

mhvk commented:
@taldcroft - I worry we will be drawing out this PR too much; I think it would be better to do it as follow up.

More generally, it seems one would typically only want to change the formatting of one or two columns. If so, maybe a `dict` would be more logical? But that might suggest not having the list option either. Anyway, my sense would be to merge this now and see before the next point release whether we want to revise anything.

taldcroft commented:
@mhvk - fair enough about getting this through.

In the case of changing just one or two columns, perhaps setting the column `format` attribute would be the preferred idiom?  At least it seems to me that if you want to print or write a table one time with a particular formatting, then most likely you'd like to have the same formatting for subsequent operations.

Sure, we can allow a dict as well as a list, but then we end up with 3 ways to do (roughly) the same thing (granted, the particular use cases will be different).

agroener commented:
@taldcroft I like the idea in principle. However, for changing alignment of multiple columns simultaneously, I don't see any major need to do it that way rather than with a list. The one exception to this I think, which you've already pointed out, is using a single string character to update the alignment of ALL columns. I stayed away from this for the time being because I think what we have now works well enough. Perhaps in a future PR, though.

agroener commented:
Okay - so more progress, but I'm not sure what this last error is all about in 'continuous-integration/travis-ci/pr' is:

    No output has been received in the last ... minutes, this potentially indicates a stalled build or something wrong with the build itself.
    The build has been terminated

taldcroft commented:
@agroener - I just restarted the job.  Sometimes jobs fail for reasons unrelated to the testing.

agroener commented:
@taldcroft Awesome, thanks! I think this is good to go.

taldcroft commented:
@astrofrog or @embray (or anyone) - this PR shows a diff for astropy-helpers.  Does @agroener need to rebase on the latest master, or is it OK to merge as is?

astrofrog commented:
@taldcroft - unfortunately this requires a rebase - we should never have changes in astropy-helpers unless in a dedicated pull request. If you prefer, you can just rebase and merge manually.

taldcroft commented:
@agroener - as you've seen there is one final issue that needs to be resolved.. you need to rebase on master to get rid of the diff on astropy-helpers.  You can do:
```
git checkout master
git pull upstream master  # assuming you've named the main astropy repo "upstream"
git checkout amg/table-format-alignment
git submodule update  # don't know if this is required but it can't hurt
git rebase master
git diff master --stat  # make sure there is no diff for astropy-helpers
git push -f origin amg/table-format-alignment
```

agroener commented:
Done. Hopefully I did it correctly. After pulling from upstream, it required a merge (and me having to go through the changes in pprint.py and table.py).

astrofrog commented:
@agroener - I think something might have gone wrong, since there should be no merge commit in this PR. I can provide instructions for rebasing to get rid of the last two commits tomorrow.

taldcroft commented:
@agroener - I just put in PR #3644 which fixes the submodule issue and we can merge that.  You are still the author of all the commits and get credit for this enhancement.

@astrofrog - I learned a little more about submodules just now.  I'm not sure if there is a more elegant way to fix up a mismatch but I just added a final commit which restores the astropy_helpers subrepo to the correct commit.

agroener commented:
@taldcroft Fair enough! Thanks for all of the help on this.

mhvk commented:
@agroener - happy to see this merged; thanks for your work!

agroener commented:
@mhvk Thanks! Couldn't have done it without you and everyone else :)
